flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/matrix/SparseVector.java

```diff
@@ -0,0 +1,791 @@
+	}
+
+	@Override
+	public double normL1() {
+		double d = 0;
+		for (double t : values) {
+			d += Math.abs(t);
+		}
+		return d;
+	}
+
+	@Override
+	public double normInf() {
+		double d = 0;
+		for (double t : values) {
+			d = Math.max(Math.abs(t), d);
+		}
+		return d;
+	}
+
+	@Override
+	public double normL2Square() {
+		double d = 0;
+		for (double t : values) {
+			d += t * t;
+		}
+		return d;
+	}
+
+	@Override
+	public SparseVector slice(int[] indices) {
+		SparseVector sliced = new SparseVector(indices.length);
+		int nnz = 0;
+		sliced.indices = new int[indices.length];
+		sliced.values = new double[indices.length];
+
+		for (int i = 0; i < indices.length; i++) {
+			int pos = Arrays.binarySearch(this.indices, indices[i]);
+			if (pos >= 0) {
+				sliced.indices[nnz] = i;
+				sliced.values[nnz] = this.values[pos];
+				nnz++;
+			}
+		}
+
+		if (nnz < sliced.indices.length) {
+			sliced.indices = Arrays.copyOf(sliced.indices, nnz);
+			sliced.values = Arrays.copyOf(sliced.values, nnz);
+		}
+
+		return sliced;
+	}
+
+	public SparseVector plus(SparseVector other) {
+		if (this.size() != other.size()) {
+			throw new RuntimeException("The size of the two vectors are different.");
+		}
+
+		int totNnz = this.values.length + other.values.length;
+		int p0 = 0;
+		int p1 = 0;
+		while (p0 < this.values.length && p1 < other.values.length) {
+			if (this.indices[p0] == other.indices[p1]) {
+				totNnz--;
+				p0++;
+				p1++;
+			} else if (this.indices[p0] < other.indices[p1]) {
+				p0++;
+			} else {
+				p1++;
+			}
+		}
+
+		SparseVector r = new SparseVector(this.size());
+		r.indices = new int[totNnz];
+		r.values = new double[totNnz];
+		p0 = p1 = 0;
+		int pos = 0;
+		while (pos < totNnz) {
+			if (p0 < this.values.length && p1 < other.values.length) {
+				if (this.indices[p0] == other.indices[p1]) {
+					r.indices[pos] = this.indices[p0];
+					r.values[pos] = this.values[p0] + other.values[p1];
+					p0++;
+					p1++;
+				} else if (this.indices[p0] < other.indices[p1]) {
+					r.indices[pos] = this.indices[p0];
+					r.values[pos] = this.values[p0];
+					p0++;
+				} else {
+					r.indices[pos] = other.indices[p1];
+					r.values[pos] = other.values[p1];
+					p1++;
+				}
+				pos++;
+			} else {
+				if (p0 < this.values.length) {
+					r.indices[pos] = this.indices[p0];
+					r.values[pos] = this.values[p0];
+					p0++;
+					pos++;
+					continue;
+				}
+				if (p1 < other.values.length) {
+					r.indices[pos] = other.indices[p1];
+					r.values[pos] = other.values[p1];
+					p1++;
+					pos++;
+					continue;
+				}
+			}
+		}
+		return r;
+	}
+
+	public DenseVector plus(DenseVector other) {
+		if (this.size() != other.size()) {
+			throw new RuntimeException("The size of the two vectors are different.");
+		}
+
+		DenseVector r = other.clone();
+		for (int i = 0; i < this.indices.length; i++) {
+			r.add(this.indices[i], this.values[i]);
+		}
+		return r;
+	}
+
+	@Override
+	public Vector plus(Vector vec) {
```

Comment:
I am wondering if this method will ever be invoked because there are two methods with more specific signature... Ideally we should only have one plus method interact with Vector and return a specific vector type. So having three subtly different `plus` and `minus` method seems a little awkward. But that is probably not a big deal at the moment.

Reply:
Indeed the three methods are confusing, so we keep only one.